optimade/filterparser/tests/test_filterparser.py

```diff
@@ -38,3 +38,166 @@ def test_repr(self):
+        self.assertIsInstance(self.parse('_exmpl_element_counts HAS < 3 AND '
+                                         '_exmpl_element_counts HAS ANY > 3, = 6, 4, != 8'), Tree)
+        self.assertIsInstance(self.parse('elements:_exmpl_element_counts:_exmpl_element_weights '
+                                         'HAS ANY > 3:"He":>55.3 , = 6:>"Ti":<37.6 , 8:<"Ga":0'), Tree)
+
+    def test_properties(self):
+        #  Filtering on Properties with unknown value
+        self.assertIsInstance(self.parse('chemical_formula_hill IS KNOWN AND '
+                                         'NOT chemical_formula_anonymous IS UNKNOWN'), Tree)
+
+    def test_precedence(self):
+        self.assertIsInstance(self.parse('NOT a > b OR c = 100 AND f = "C2 H6"'), Tree)
+        self.assertIsInstance(self.parse('(NOT (a > b)) OR ( (c = 100) AND (f = "C2 H6") )'), Tree)
+        self.assertIsInstance(self.parse('a >= 0 AND NOT b < c OR c = 0'), Tree)
+        self.assertIsInstance(self.parse('((a >= 0) AND (NOT (b < c))) OR (c = 0)'), Tree)
+
+    def test_special_cases(self):
+        self.assertIsInstance(self.parse('te < st'), Tree)
+        self.assertIsInstance(self.parse('spacegroup="P2"'), Tree)
+        self.assertIsInstance(self.parse('_cod_cell_volume<100.0'), Tree)
+        self.assertIsInstance(self.parse('_mp_bandgap > 5.0 AND _cod_molecular_weight < 350'), Tree)
+        self.assertIsInstance(self.parse('_cod_melting_point<300 AND nelements=4 AND elements="Si,O2"'), Tree)
+        self.assertIsInstance(self.parse('key=value'), Tree)
+        self.assertIsInstance(self.parse('author=" someone "'), Tree)
+        self.assertIsInstance(self.parse('author=" som\neone "'), Tree)
+        self.assertIsInstance(self.parse('number=0.ANDnumber=.0ANDnumber=0.0ANDnumber=+0AND_n_u_m_b_e_r_=-0AND'
+                                         'number=0e1ANDnumber=0e-1ANDnumber=0e+1'), Tree)
+
+        self.assertIsInstance(self.parse('NOTice=val'), Tree)  # property (ice) != property (val)
+        self.assertIsInstance(self.parse('NOTice="val"'), Tree)  # property (ice) != value ("val")
+        self.assertIsInstance(self.parse('"NOTice"=val'), Tree)  # value ("NOTice") = property (val)
+
+        with self.assertRaises(ParserError):
+            self.parse('NOTICE=val')  # not valid property or value (NOTICE)
+        with self.assertRaises(ParserError):
+            self.parse('"NOTICE"=Val')  # not valid property (Val)
+        with self.assertRaises(ParserError):
+            self.parse('NOTICE=val')  # not valid property or value (NOTICE)
+
+    def test_parser_version(self):
+        self.assertEqual(self.parser.version, self.version)
+        self.assertEqual(self.parser.variant, self.variant)
+
+    def test_repr(self):
+        self.assertIsNotNone(repr(self.parser))
+        self.parser.parse('key="value"')
+        self.assertIsNotNone(repr(self.parser))
+
+
+if __name__ == '__main__':
```

Comment:
I don't know that this is needed for pytest?

Reply:
I just used it to run the test locally by running the script itself. I will remove it.